### PR TITLE
Ven 58647 fix

### PR DESCRIPTION
**Pull request by rvelaVenafi**
VEN-58647.
Implemented, Transition from Username/Password to Token Auth for TPP.
Added test cases for TPP token authorization.

Requesting @asquith-venafi and @tr1ck3r to review this pull request as it contains changes from both @angelmoo and @rvelaVenafi

### CODE REVIEW

File: venafi/provider.go [hunk 1]
```diff
@@ -45,12 +45,20 @@ Example for Venafi Cloud: Default`,
...
 				Type:        schema.TypeString,
 				Optional:    true,
 				DefaultFunc: schema.EnvDefaultFunc("VENAFI_PASS", nil),
 				Description: `Password for WebSDK user. Example: password`,
+				Deprecated:  ", please use access_token instead",
+			},
+			"access_token": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("VENAFI_ACCESS", nil),
```
Comment (tr1ck3r) [inline venafi/provider.go]: Environment variable should be called "VENAFI_TOKEN"

File: venafi/resource_venafi_certificate_test.go [hunk 2]
```diff
@@ -403,15 +446,157 @@ func TestTPPECDSASignedCert(t *testing.T) {
...
+	domain := "venafi.example.com"
+	data.cn = rand + "." + domain
+	data.dns_ns = "alt-" + data.cn
+	data.dns_ip = "192.168.1.1"
+	data.dns_email = "venafi@example.com"
+	data.private_key_password = "123xxx"
+	data.key_algo = ecdsa521
+	data.expiration_window = 168
+	config := fmt.Sprintf(tokenConfig, tokenProviderECDSA, data.cn, data.dns_ns, data.dns_ip, data.dns_email, data.key_algo, data.private_key_password, data.expiration_window)
+	t.Logf("Testing TPP Token certificate with ECDSA key  with config:\n %s", config)
```
Comment (tr1ck3r) [inline venafi/resource_venafi_certificate_test.go]: two spaces between "key" and "with"

File: main.tf [hunk 3]
```diff
@@ -4,29 +4,43 @@ This is an example Terraform file to show capabilities of the VCert integration.
...
 
 variable "TRUST_BUNDLE" {
+  default = ""
+}
+
+variable "ACCESS_TOKEN" {
+  default = ""
+}
+
+variable "TPP_TOKEN_URL"{
```
Comment (tr1ck3r) [inline main.tf]: Do we need this?  This update only involves using a token so we don't need to communicate with a different URL (i.e. /VEDAuth).

**Review by tr1ck3r — CHANGES_REQUESTED** [review]
@rvelaVenafi looks good... a couple small things to look at

File: venafi/provider.go (hunk 1 above)
Comment (rvelaVenafi) [reply venafi/provider.go]: changed

File: venafi/resource_venafi_certificate_test.go (hunk 2 above)
Comment (rvelaVenafi) [reply venafi/resource_venafi_certificate_test.go]: removed one blank space

File: main.tf (hunk 3 above)
Comment (rvelaVenafi) [reply main.tf]: Removed unnecessary variable TPP_TOKEN_URL

**Review by tr1ck3r — APPROVED** [review]

File: venafi/resource_venafi_certificate_test.go [hunk 4]
```diff
@@ -403,15 +446,157 @@ func TestTPPECDSASignedCert(t *testing.T) {
...
 			r.TestStep{
 				Config: config,
 				Check: func(s *terraform.State) error {
 					t.Log("Testing TPP certificate second run")
 					gotSerial := data.serial
-					err := checkStandartCert(t, &data, s)
+					err := checkStandardCert(t, &data, s)
+					if err != nil {
+						return err
+					} else {
```
Comment (lhdomenech) [inline venafi/resource_venafi_certificate_test.go:460]: using else following a return is bad practices see https://golang.org/doc/effective_go.html#if

File: venafi/resource_venafi_certificate_test.go [hunk 5]
```diff
@@ -403,15 +446,157 @@ func TestTPPECDSASignedCert(t *testing.T) {
...
+						t.Logf("Compare certificate serial %s with serial after second run %s", gotSerial, data.serial)
+						if gotSerial != data.serial {
+							return fmt.Errorf("serial number from second run %s is different as in original number %s."+
+								" Which means that certificate was updated without reason", data.serial, gotSerial)
+						} else {
```
Comment (lhdomenech) [inline venafi/resource_venafi_certificate_test.go:465]: same here

File: venafi/resource_venafi_certificate_test.go [hunk 6]
```diff
@@ -403,15 +446,157 @@ func TestTPPECDSASignedCert(t *testing.T) {
...
+			},
+			r.TestStep{
+				Config: config,
+				Check: func(s *terraform.State) error {
+					t.Log("Testing TPP Token certificate update")
+					gotSerial := data.serial
+					err := checkStandardCert(t, &data, s)
+					if err != nil {
+						return err
+					} else {
```
Comment (lhdomenech) [inline venafi/resource_venafi_certificate_test.go:509]: and here

File: venafi/resource_venafi_certificate_test.go [hunk 7]
```diff
@@ -403,15 +446,157 @@ func TestTPPECDSASignedCert(t *testing.T) {
...
+						t.Logf("Compare updated original certificate serial %s with updated %s", gotSerial, data.serial)
+						if gotSerial == data.serial {
+							return fmt.Errorf("serial number from updated certificate %s is the same as "+
+								"in original number %s", data.serial, gotSerial)
+						} else {
```
Comment (lhdomenech) [inline venafi/resource_venafi_certificate_test.go:514]: and here

File: venafi/resource_venafi_certificate_test.go [hunk 8]
```diff
@@ -403,15 +446,157 @@ func TestTPPECDSASignedCert(t *testing.T) {
...
+			},
+			r.TestStep{
+				Config: config,
+				Check: func(s *terraform.State) error {
+					t.Log("Testing TPP certificate second run")
+					gotSerial := data.serial
+					err := checkStandardCert(t, &data, s)
+					if err != nil {
+						return err
+					} else {
```
Comment (lhdomenech) [inline venafi/resource_venafi_certificate_test.go:556]: and here

File: venafi/resource_venafi_certificate_test.go [hunk 9]
```diff
@@ -403,15 +446,157 @@ func TestTPPECDSASignedCert(t *testing.T) {
...
+						t.Logf("Compare certificate serial %s with serial after second run %s", gotSerial, data.serial)
+						if gotSerial != data.serial {
+							return fmt.Errorf("serial number from second run %s is different as in original number %s."+
+								" Which means that certificate was updated without reason", data.serial, gotSerial)
+						} else {
```
Comment (lhdomenech) [inline venafi/resource_venafi_certificate_test.go:561]: and here

**Review by lhdomenech — COMMENTED** [review]
change looks good, just essentially one bad practice comment, so since I don't know what best practices and linters are used, feel free to ignore this comment even more because I saw the same pattern in existing code

**Review by asquith-venafi — APPROVED** [review]

**Review by lhdomenech — APPROVED** [review]
Approving this PR since comments are not specific to new code. I created an issue for that https://github.com/Venafi/terraform-provider-venafi/issues/26

**Pull request merged**